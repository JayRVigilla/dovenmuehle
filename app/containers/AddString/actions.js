import { prependString } from '../../ApiCalls';

export default async function sendPostToAPI(string) {
  return async dispatch => {
    const response = prependString(string);
    return dispatch(addString(response.data));
  };
}

function addString(string) {
  return {
    type: 'ADD_STRING',
    payload: string,
  };
}
