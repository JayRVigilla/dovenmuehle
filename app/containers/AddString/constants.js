/*
 * AddString Constants
 * Each action has a corresponding type, which the reducer knows and picks up on.
 * To avoid weird typos between the reducer and the actions, we save them as
 * constants here. We prefix them with 'yourproject/YourComponent' so we avoid
 * reducers accidentally picking up actions they shouldn't.
 *
 * Follow this format:
 * export const YOUR_ACTION_CONSTANT = 'yourproject/YourContainer/YOUR_ACTION_CONSTANT';
 */

export const POST_STRING = 'dovenmuehle/AddString/POST_STRING';
export const PREPENDED_STRING = 'dovenmuehle/AddString/PREPENDED_STRINGS';
export const POST_STRING_ERR = 'dovenmuehle/AddString/POST_STRINGS_ERR';
export const UPDATE_CLIENT_STRING = 'dovenmuehle/AddString/CLIENT_STRING';
