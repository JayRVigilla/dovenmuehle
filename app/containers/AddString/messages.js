/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage container.
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.containers.HomePage';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'Add to your mystery plot line',
  },
  subHead: {
    id: `${scope}.subHead`,
    defaultMessage: 'Welcome to the Prepender',
  },
});
