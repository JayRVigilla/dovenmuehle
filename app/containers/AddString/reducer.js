/*
 *
 * AddString reducer
 *
 */

/* eslint-disable default-case, no-param-reassign */
import produce from 'immer';
import {
  CLIENT_STRING,
  POST_STRING,
  PREPENDED_STRING,
  POST_STRING_ERR,
} from './constants';

export const INITIAL_STATE = {
  clientString: '',
  err: false,
  prepended: false,
  isLoading: false,
};

const addStringReducer = (state = INITIAL_STATE, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CLIENT_STRING:
        draft.clientString = action.clientString;
        break;

      case POST_STRING:
        draft.isLoading = true;
        break;

      case PREPENDED_STRING:
        draft.isLoading = false;
        draft.prepended = true;
        break;

      case POST_STRING_ERR:
        draft.prepended = false;
        draft.err = action.err;
        break;
    }
  });

export default addStringReducer;
