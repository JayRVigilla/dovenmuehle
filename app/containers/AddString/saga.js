import { takeLeading, call } from 'redux-saga/effects';
import { prependString } from '../../ApiCalls';

function* addStringAsync() {
  // yield put({ type: 'ADD_STRING_ASYNC', payload: str });
  // console.log('running AddStringAsync');
  yield call(prependString);
  // const result = yield prependString(str);
}

export default function* rootSaga() {
  // console.log('running AddString rootSaga()');
  yield takeLeading('ADD_STRING', addStringAsync);
  // yield takeLeading('ADD_STRING', prependString);
}
