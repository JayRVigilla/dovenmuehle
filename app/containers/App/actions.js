import { GET_STRINGS, STORE_STRINGS, GET_STRINGS_ERR } from './constants';

export function getStringsArray() {
  return {
    type: GET_STRINGS,
    isLoading: true,
    err: false,
    strings: false,
  };
}

export function storeStrings(strings) {
  return {
    type: STORE_STRINGS,
    strings,
    isLoading: false,
  };
}

export function errorStrings(err) {
  return {
    type: GET_STRINGS_ERR,
    err,
    isLoading: false,
  };
}
