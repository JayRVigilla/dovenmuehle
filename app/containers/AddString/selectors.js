import { createSelector } from 'reselect';
import { INITIAL_STATE } from './reducer';

const selectAddString = state => state.addString || INITIAL_STATE;

const makeSelectStrings = () =>
  createSelector(
    selectAddString,
    addStringState => addStringState.strings,
  );

const makeSelectIsLoading = () =>
  createSelector(
    selectAddString,
    addStringState => addStringState.isLoading,
  );

const makeSelectErr = () =>
  createSelector(
    selectAddString,
    addStringState => addStringState.err,
  );

export {
  selectAddString,
  makeSelectStrings,
  makeSelectIsLoading,
  makeSelectErr,
};
