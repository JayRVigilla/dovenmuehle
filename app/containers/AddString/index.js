/*
 * AddString
 *
 * form that allows client to add to txt file in backend
 *
 */

import React from 'react';
// import { useHistory } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import { compose } from 'redux';
import { connect } from 'react-redux';
import injectSaga from 'utils/injectSaga';
import useInjectReducer from 'utils/injectReducer';
import { createStructuredSelector } from 'reselect';
import saga from './saga';
import messages from './messages';
// import LoadingIndicator from '../../components/LoadingIndicator';
import {
  makeSelectClientString,
  makeSelectIsLoading,
  makeSelectErr,
} from './selectors';
import { postString, updateClientString, postSringsError } from './actions';
import reducer from './reducer';

const key = 'addString';

const mapStateToProps = createStructuredSelector({
  clientString: makeSelectClientString(),
  isLoading: makeSelectIsLoading(),
  err: makeSelectErr(),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = useInjectReducer({ key, reducer });

const withSaga = injectSaga({
  key,
  saga,
});

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(function AddString({
  clientString,
  // isLoading,
  // prepended, if false then pop up message "cannot be all whitespace"
  onSubmit,
  onUpdateClientString,
}) {
  // use effect?

  return (
    <div>
      <h1>
        <FormattedMessage {...messages.header} />
      </h1>

      <div>
        <h2>
          <FormattedMessage {...messages.subHead} />
        </h2>
        <p>To add a event towards the beginning of the story:</p>
        <ul>
          <li>Enter your plot point</li>
          <li>Click Submit</li>
        </ul>
        <a href="/">See your mystery</a>
      </div>

      <div>
        <form onSubmit={onSubmit}>
          <label htmlFor="clientString">
            <input
              id="clientString"
              type="text"
              value={clientString}
              onChange={onUpdateClientString}
            />
          </label>
          <button type="submit">Submit</button>
        </form>
      </div>
    </div>
  );
});

export function mapDispatchToProps(dispatch) {
  return {
    onUpdateClientString: evt => dispatch(updateClientString(evt.target.value)),
    onSubmit: evt => {
      if (
        evt !== undefined &&
        evt.preventDefault &&
        // regex checks if at least one character is not whitespace
        // pulled from https://stackoverflow.com/questions/2249460/how-to-use-javascript-regex-to-check-for-empty-form-fields
        /([^\s])/.test(evt.target.value)
      ) {
        evt.preventDefault();
        dispatch(postString());
        dispatch(updateClientString(''));
      }
      dispatch(postSringsError());
    },
  };
}
