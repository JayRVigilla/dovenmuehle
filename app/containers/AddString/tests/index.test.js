import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import AddString, { mapDispatchToProps } from '../index';
import { updateClientString, postString } from '../actions';
import configureStore from '../../../configureStore';

describe('<AddString />', () => {
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <AddString />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should should go to HomePage when link clicked', () => {
    // trigger link click
    expect().toSomething();
  });
  describe('mapDispatchtoProps', () => {
    describe('onUpdateClientstring', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onUpdateClientString).toBeDefined();
      });

      it('should dispatch onUpdateClientString when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const testString = 'mxstbr';
        result.onUpdateClientString({ target: { value: testString } });
        expect(dispatch).toHaveBeenCalledWith(updateClientString(testString));
      });

      // will have to mock dispatches... probably
      describe('onSubmit', () => {
        it('should be injected', () => {
          const dispatch = jest.fn;
          const result = mapDispatchToProps(dispatch);
          expect(result.onSubmit).toBeDefined();
        });

        it('should dispatch onSubmit when called', () => {
          const dispatch = jest.fn();
          const result = mapDispatchToProps(dispatch);
          result.onSubmit();
          expect(dispatch).toHaveBeenCalledWith(postString());
        });
      });
    });

    it('', () => {});
  });
});
