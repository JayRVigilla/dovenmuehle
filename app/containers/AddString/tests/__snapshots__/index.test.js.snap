// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddString /> should render and match the snapshot 1`] = `
<div>
  <h1>
    <span>
      Add to your mystery plot line
    </span>
  </h1>
  <div>
    <h2>
      <span>
        Welcome to the Prepender
      </span>
    </h2>
    <p>
      To add a event towards the beginning of the story:
    </p>
    <ul>
      <li>
        Enter your plot point
      </li>
      <li>
        Click Submit
      </li>
    </ul>
    <a
      href="/"
    >
      See your mystery
    </a>
  </div>
  <div>
    <form>
      <label
        for="clientString"
      >
        <input
          id="clientString"
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
